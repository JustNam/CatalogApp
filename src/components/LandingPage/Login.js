import React, { Component } from 'react';
import { Redirect, Link } from 'react-router-dom';
import { Button } from 'react-bootstrap';

class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      password: '',
      redirect: false,
      error: false,
    };
  }

  handleUsernameChange = (e) => {
    this.setState({ username: e.target.value });
  };

  handlePasswordChange = (e) => {
    this.setState({ password: e.target.value });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { username, password } = this.state;
    this.props.login(username, password).then((response) => {
      if (response.payload.error) {
        this.setState({ error: true });
      } else {
        this.setState({ redirect: true });
      }
    });
  };

  render() {
    const { username, password, redirect, error } = this.state;
    if (redirect) {
      return <Redirect to="/categories" />;
    }
    return (
      <div className="container">
        <div className="row">
          <div className="col-sm-9 col-md-7 col-lg-5 mx-auto">
            <div className="card card-signin my-5">
              <div className="card-body">
                <h5 className="card-username text-center">Catalog App</h5>
                <form className="form-signin">
                  {this.props.location.signUpSuccess && (
                    <div className="success-message">
                      Sign up successfully, log in with your new account!
                    </div>
                  )}
                  {error && (
                    <div className="invalid-credential">
                      Invalid username or password
                    </div>
                  )}
                  <div className="form-label-group">
                    <input
                      type="text"
                      id="inputUsername"
                      className="form-control"
                      placeholder="Email address"
                      value={username}
                      onChange={this.handleUsernameChange}
                      required
                    />
                    <label htmlFor="inputUsername">Username</label>
                  </div>

                  <div className="form-label-group">
                    <input
                      type="password"
                      id="inputPassword"
                      className="form-control"
                      placeholder="Password"
                      value={password}
                      onChange={this.handlePasswordChange}
                      required
                    />
                    <label htmlFor="inputPassword">Password</label>
                  </div>

                  <button
                    className="btn btn-lg btn-primary btn-block text-uppercase"
                    onClick={this.handleSubmit}
                    type="submit"
                  >
                    Log in
                  </button>
                  <Button
                    href="/signup"
                    className="btn btn-lg btn-outline-secondary btn-block text-uppercase"
                    type="button"
                  >
                    Sign up
                  </Button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
export default LoginPage;
