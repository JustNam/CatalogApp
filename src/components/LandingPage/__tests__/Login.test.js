import React from 'react';
import { shallow } from 'enzyme';
import Login from '../Login';

describe('Login with correct credentials', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      location: {
        signUpSuccess: false,
      },
      login: jest.fn(() => Promise.resolve({
        payload: {
          data: '',
        },
      })),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<Login {...props} />);
    instance = wrapper.instance();
  });
  it('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('It should change username in state when username input is changed', () => {
    instance.handleUsernameChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().username).toBe('test');
  });
  it('It should change password in state when password input is changed', () => {
    instance.handlePasswordChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().password).toBe('test');
  });
  it('It should redirect if the request is valid', async () => {
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    await expect(wrapper.state().redirect).toBe(true);
  });
});

describe('Login with input validations', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      location: {
        signUpSuccess: true,
      },
      login: jest.fn(() => Promise.resolve({
        payload: {
          error: 'test',
        },
      })),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<Login {...props} />);
    instance = wrapper.instance();
  });
  it('It should show error when is request is invalid', async () => {
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    await expect(wrapper.state().error).toBe(true);
  });
});