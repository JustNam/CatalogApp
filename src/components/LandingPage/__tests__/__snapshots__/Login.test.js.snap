// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login with correct credentials It should render correctly 1`] = `ShallowWrapper {}`;
